src/cargo/ops/tree/graph.rs

```diff
@@ -233,9 +233,35 @@ impl<'a> Graph<'a> {
 
         let mut dupes: Vec<(&Node, usize)> = packages
             .into_iter()
-            .filter(|(_name, indexes)| indexes.len() > 1)
+            .filter(|(_name, indexes)| {
+                let mut seen_pkg = HashSet::new();
+                indexes
+                    .into_iter()
+                    .filter(|(node, _)| {
+                        match node {
+                            Node::Package {
+                                package_id,
+                                features,
+                                ..
+                            } => {
+                                // Do not treat duplicates on the host or target as duplicates.
+                                let ignore_kind_package = Node::Package {
+                                    package_id: package_id.clone(),
+                                    features: features.clone(),
+                                    kind: CompileKind::Host,
+                                };
+                                !seen_pkg.contains(&ignore_kind_package)
+                                    && seen_pkg.insert(ignore_kind_package)
```

Comment:
Sorry I didn't say it clear in https://github.com/rust-lang/cargo/pull/10466#discussion_r837577088. What do you think about dropping `contains` and using only `insert` operation here? This is only a little improvement and might not affect the performance but to me it's eye-pleasing 😆 

```suggestion
                                seen_pkg.insert(ignore_kind_package)
```

Reply:
I wonder if this would make this code less readable? Maybe determining if it exists is a better indication that we need to de-dupe?

Reply:
Addressed!

Reply:
Oh I just noticed that we can push it further by collecting to a `HashSet`:

```rust
indexes.into_iter().map(....).collect::<HashSet<_>>().len() > 1
```

Sorry about the back-and-forth, but this actually looks better and is a common dedup pattern I guess.